### DISCUSSION TITLE
How to hide completion codes

### DISCUSSION THREAD

**Discussion by kurokida** [jspsych/jsPsych#2759] [Q&A] [answered]
Hi all,

I've read the following two discussions, but my case is slightly different from them.
https://github.com/jspsych/jsPsych/discussions/1170
https://github.com/jspsych/jsPsych/discussions/710

So far I have been writing a fixed completion code directly in the HTML file, but if participants check the source file they can easily find this out. Is there a way to hide the code?

The code does not need to be different for each participant, it is fixed. Rather it should be fixed for my experiment.
I'm not familiar with PHP programming, but it could be accomplished by creating a PHP file that returns the code because the content of the PHP file is invisible.

Do you have any ideas?　If my PHP idea works, it would be very helpful if you could show me some sample code.

**Comment by kurokida** [comment, marked as answer]
I've solved this problem by myself, but I'm not sure if this approach is correct.
If you notice anything let me know.

I made a very simple PHP file like this as getCompletionCode.php.

```php
<?php
    echo '1234'; // This is the fixed code
?>
```

The following code is written in the experiment file.

```javascript
    let completion_code;
    const getCode = {
        type: jsPsychCallFunction,
        async: true,
        func: function(done){
            console.log('getCompletionCode')
            const xhr = new XMLHttpRequest();
            xhr.open('POST', 'getCompletionCode.php');
            xhr.setRequestHeader('Content-Type', 'application/json');
            xhr.onload = function() {
                if(xhr.status == 200){
                    completion_code = JSON.parse(xhr.responseText);
                    console.log(completion_code);
                }
                done(); // invoking done() causes experiment to progress to next trial.
            };

            xhr.send(null); 

        }
    }
```

I can receive the code and the code can not be probably seen from the source file.

**Reply by jodeleeuw** [reply]
Thanks for sharing, @kurokida !

I'm guessing you are aware of this, but just for others who are reading this looking for a solution: This doesn't fully prevent someone from accessing the completion code before completing the experiment. Someone could manually run the `xhr` portion of the code to get the completion code. I don't know of any way to fully obfuscate a completion code and be able to display it to the user in cleartext at some point during the experiment.

**Reply by kurokida** [reply]
Thank you for your reply.
Actually, I haven't noticed the problem my solution has. Now I totally understood it.
Although my solution seems to be better than directly writing the completion codes in the file, the benefit is very limited.
